### PR TITLE
Force Two Factor

### PR DESCRIPTION
**Pull request by mikeselander**
Adds a network-level setting to require two-factor authentication on a per-role or universal level. If 2fa forcing is on and a user matches the criteria, we force the user into a 2fa settings screen and they cannot use the site until they add valid 2fa to their account.

**Super administrators can edit the settings with this interface via network settings:**
screen shot 2018-09-04 at 1 40 41 pm

**Users who fall within the required buckets will see this interface:**
screen shot 2018-09-04 at 1 42 16 pm

### CODE REVIEW

**Comment by mikeselander** [comment]
@joehoyle can I please get a review and test from you on this functionality?

It's specifically setup for networks at the moment and also worth noting that I've written the code inline with the standards and style setup for the upstream repo. A lot of the style and architectural pattern is pretty gross, but hoping that we can get this upstreamed faster by going that route.

**Comment by joehoyle** [comment]
@mikeselander functinalty wise: it's working great! Though the enable screen looks like this:



Have those items at the bottom shown. I ran `npm install` and `grunt`, and then I don't have any errors in console for missing assets. Also if I submit it and there's an error, it seems it doesn't show any validation.

+1 for a non-multisite support too, as `central.aws.hmn.md` isn't multisite. I'd imagine you could just use the existing functions for that?

**Comment by mikeselander** [comment]
> Have those items at the bottom shown.

Odd, I have never seen those 🤔 I'll try to re-produce.

> Also if I submit it and there's an error, it seems it doesn't show any validation.

This is an architectural problem with the plugin itself that would require modifying a good bit of internals to fix. The plugin is built to only ever work on a save submission from a user panel and none of them return or throw errors :(

Some of the methods (TOTP for example) do show an error message within their little pane if something has gone wrong.

> +1 for a non-multisite support too, as central.aws.hmn.md isn't multisite. I'd imagine you could just use the existing functions for that?

Ya, I can mostly use the same functions and then registering the options for single sites. I'll add that in 👍

**Comment by mikeselander** [comment]
@joehoyle I've fixed this visual issues and added single site support. Mind taking another look through?

**Comment by rmccue** [comment]
Should this be PR'd directly to the upstream repo rather than merged?

**Comment by joehoyle** [comment]
@mikeselander nice, it's looking pretty good to me. I don't know why the plugin shows the "Security Keys" section if you haven't selected U2F.

Out of interest, does the plugin allow you to filter the options for 2fa? I guess it could be a site setting again, but it seems odd site administrators can't whitelist the allowed 2fa authentication methods.

File: class.two-factor-core.php [hunk 1]
```diff
@@ -697,4 +860,247 @@ public static function user_two_factor_options_update( $user_id ) {
...
+		// If we can't pull up user, escape.
+		if ( ! ( $user instanceof WP_User ) ) {
+			return false;
+		}
+
+		// Check whether a user is in a user role that requires two-factor authentication.
+		$two_factor_forced_roles = self::get_forced_user_roles();
+		$required_roles          = array_filter( $user->roles, function( $role ) use ( $two_factor_forced_roles ) {
+			return in_array( $role, $two_factor_forced_roles, true );
+		}, ARRAY_FILTER_USE_BOTH);
```
Comment (joehoyle) [inline class.two-factor-core.php]: Missing space at the end.

File: class.two-factor-core.php [hunk 2]
```diff
@@ -697,4 +860,247 @@ public static function user_two_factor_options_update( $user_id ) {
...
+
+	/**
+	 * Get whether site has two-factor universally forced or not.
+	 *
+	 * @since 0.1-dev
+	 *
+	 * @return bool
+	 */
+	public static function get_universally_forced_option() {
+		$is_forced = is_multisite() ? get_site_option( self::FORCED_SITE_META_KEY, false ) : get_option( self::FORCED_SITE_META_KEY, false );
```
Comment (joehoyle) [inline class.two-factor-core.php]: You don't need this ternary:

> This function is almost identical to get_option(), except that in multisite, it returns the network-wide option. For non-multisite installs, it uses get_option.

File: class.two-factor-core.php [hunk 3]
```diff
@@ -697,4 +860,247 @@ public static function user_two_factor_options_update( $user_id ) {
...
+	 */
+	public static function force_two_factor_setting_options() {
+		?>
+		<h2><?php esc_html_e( 'Two-Factor Options', 'two-factor' ); ?></h2>
+		<table class="form-table">
+			<?php wp_nonce_field( 'force_two_factor_options', '_nonce_force_two_factor_options', false ); ?>
+			<tbody>
+				<tr>
+					<th scope="row">
+						<?php esc_html_e( 'Universally force two-factor', 'two-factor' ); ?>
```
Comment (joehoyle) [inline class.two-factor-core.php]: I'm not sure of this working. "Force all users to enable two-factor"?

**Comment by joehoyle** [comment]
@mikeselander actually, I see for enable 2fa you are doing it inline, as opposed to redirecting to a specific URL. For whatever reason this means if I am on hte homepage, it shows like this: 



It might be better to have a discrete  URL where this is shown (maybe wp-login.php?action=force-2fa for example), rather than trying to show it in-place on all URLs.

**Comment by joehoyle** [comment]
Also noticed I have a fatal on the homepage screenshot:

```Fatal error: Uncaught Error: Call to undefined function convert_to_screen() in /usr/src/app/wordpress/wp-admin/includes/class-wp-list-table.php:132 Stack trace: #0 /usr/src/app/content/plugins/two-factor/providers/class.two-factor-fido-u2f-admin.php(189): WP_List_Table->__construct() #1 /usr/src/app/wordpress/wp-includes/class-wp-hook.php(286): Two_Factor_FIDO_U2F_Admin::show_user_profile(Object(WP_User)) #2 /usr/src/app/wordpress/wp-includes/class-wp-hook.php(310): WP_Hook->apply_filters('', Array) #3 /usr/src/app/wordpress/wp-includes/plugin.php(453): WP_Hook->do_action(Array) #4 /usr/src/app/content/plugins/two-factor/class.two-factor-core.php(827): do_action('show_user_secur...', Object(WP_User)) #5 /usr/src/app/content/plugins/two-factor/class.two-factor-core.php(537): Two_Factor_Core::user_two_factor_options(Object(WP_User)) #6 /usr/src/app/content/plugins/two-factor/class.two-factor-core.php(501): Two_Factor_Core::force_2fa_login_html() #7 /usr/src/app/wordpress/wp-includes/class-wp-hook.php(286): Two_Factor_Core:: in /usr/src/app/wordpress/wp-admin/includes/class-wp-list-table.php on line 132```

File: class.two-factor-core.php (hunk 2 above)
Comment (mikeselander) [reply class.two-factor-core.php]: Ah, TIL :)

**Comment by mikeselander** [comment]
> Also noticed I have a fatal on the homepage screenshot:

Will look and re-produce 👍 

> @mikeselander nice, it's looking pretty good to me. I don't know why the plugin shows the "Security Keys" section if you haven't selected U2F.

Ya, that's just something that it does ¯\\\_(ツ)\_/¯

> Out of interest, does the plugin allow you to filter the options for 2fa? I guess it could be a site setting again, but it seems odd site administrators can't whitelist the allowed 2fa authentication methods.

It does allow you to filter the methods that are available - The filter is `two_factor_providers` and can be found here: https://github.com/humanmade/two-factor/blob/master/class.two-factor-core.php#L77-L86

**Comment by mikeselander** [comment]
> Should this be PR'd directly to the upstream repo rather than merged?

@rmccue I wanted to get Joe's feedback before opening a PR against the upstream repo. I also am wary of how long this could take to get merged into the upstream repo and my client needs it this week.

**Comment by rmccue** [comment]
> I wanted to get Joe's feedback before opening a PR against the upstream repo. I also am wary of how long this could take to get merged into the upstream repo and my client needs it this week.

Yeah, just wanted to check if we wanted to fork or not :)

**Comment by rmccue** [comment]
Linking the upstream PR: https://github.com/georgestephanis/two-factor/pull/239

Should we close this PR in favour of handling it over there?

**Comment by shadyvb** [comment]
Opening back now, since upstream PR hasn't been merged.

**Review by shadyvb — APPROVED** [review]

**Comment by shadyvb** [comment]
Travis install seems to be busted here, so will ignore and merge, tested the functionality and it seems to be working as expected.

**Comment by roborourke** [comment]
Yeah I think it's missing a git submodule for the xwp dev library

**Comment by roborourke** [comment]
What are your thoughts on the release version? 0.2.2? technically 0.2.1 introduced a bug by missing a bunch of this plugin :/

We can go for 0.3.0 but we'll need to backport that update through to v7

**Pull request merged**